ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/SubmitterInfo.scala

```diff
@@ -3,35 +3,71 @@
 
 package com.daml.ledger.participant.state.v1
 
-import java.time.Instant
+import com.google.protobuf.ByteString
 
 /** Information provided by the submitter of changes submitted to the ledger.
   *
   * Note that this is used for party-originating changes only. They are
   * usually issued via the Ledger API.
   *
-  * @param actAs: the non-empty set of parties that submitted the change.
+  * @param actAs the non-empty set of parties that submitted the change.
   *
-  * @param applicationId: an identifier for the Daml application that
-  *   submitted the command. This is used for monitoring and to allow Daml
-  *   applications subscribe to their own submissions only.
+  * @param applicationId an identifier for the Daml application that
+  *   submitted the command. This is used for monitoring, command deduplication,
+  *   and to allow Daml applications subscribe to their own submissions only.
   *
-  * @param commandId: a submitter provided identifier that he can use to
-  *   correlate the stream of changes to the participant state with the
-  *   changes he submitted.
+  * @param commandId a submitter-provided identifier to identify an intended ledger change
+  *   within all the submissions by the same parties and application.
   *
-  * @param deduplicateUntil: the time until which the command should be deduplicated.
-  *   Command deduplication is already performed by the participant.
-  *   The ledger may choose to perform additional (cross-participant)
-  *   command deduplication. If it chooses to do so, it must follow the
-  *   same rules as the participant:
-  *   - Deduplication is based on the (submitter, commandId) tuple.
-  *   - Commands must not be deduplicated after the deduplicateUntil time has passed.
-  *   - Commands should not be deduplicated after the command was rejected.
+  * @param deduplicationPeriod The deduplication period for the command submission.
+  *   Used for the deduplication guarantee described in the [[ReadService.stateUpdates]].
+  *
+  * @param submissionId An identifier for the submission that allows an application
+  *   to correlate completions to its submissions.
+  *
+  * @param submissionRank The rank of the submission among all submissions with the same change ID.
+  *   Used for the submission rank guarantee described in the [[ReadService.stateUpdates]].
   */
 final case class SubmitterInfo(
     actAs: List[Party],
     applicationId: ApplicationId,
     commandId: CommandId,
-    deduplicateUntil: Instant,
-)
+    deduplicationPeriod: DeduplicationPeriod,
+    submissionId: SubmissionId,
+    submissionRank: Offset,
+) {
+
+  /** The ID for the ledger change */
+  val changeId: ChangeId = new ChangeId(applicationId, commandId, actAs.toSet)
+}
+
+/** Identifier for ledger changes used by command deduplication
+  *
+  * @see ReadService.stateUpdates for the command deduplication guarantee
+  */
+class ChangeId(
+    private val applicationId: ApplicationId,
```

Comment:
@andreaslochbihler-da : so far we didn't use the `applicationId` as part of the change-id. Adding it now without binding it to a version (e.g., the transaction version of the submitted transaction), will break backwards compatibility.

How important is that change? Should we just leave the `applicationId` out of the changeId?

Reply:
Our current documentation is inconsistent whether the application ID should be part of the change ID or not (https://github.com/digital-asset/daml/blob/main/ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto#L38 says that application ID should be part of it whereas https://docs.daml.com/app-dev/app-arch.html#command-deduplication doesn't mention the application ID).

As part of command dedup cleanup, we should make up our mind whether we want to provide cross-application command deduplication in the long run. If not, then now is a good time to add the application ID to the change ID.

Not providing command dedup across application IDs provides the following benefits:
* Several applications cannot interfere with each other if they happen to choose the same command ID.
* We may want to offer different `max_deduplication_time`s for different applications (high-throughput applications need higher SLAs for long dedup times).

So ultimately this comes down to a product decision. During the design discussion, @bame-da confirmed that he doesn't see a use case for command dedup across applications.

As for backwards compatibility, I don't think we need versioning. When we switch a Daml ledger to the new scheme, we just have to check that there is no deduplication entry for the old format (i.e., without application ID), but we don't need to create one in the old format (provided participants of one application switch over to the format in one step). Since the lifetime of command dedup is bounded, we can eventually completely get rid of this additional check.

Reply:
I'm all for including the `applicationId` in the `changeId`.

We do have both KV production ledgers as well as daml-on-sql production ledgers. So, we need to devise a way to roll-out these changes in a way that guarantees deduplication across upgrades. Your approach of checking the duplication keys for both the `(actAs, commandId)` and `(actAs, applicationId, commandId)` sounds good. We can then remove the check for `(actAs, commandId)` in a later release (assuming deduplication is not required for multiple weeks).